In M1 Ventura safari 16.1, when the constraints passed in by getDisplayMedia contain the 'ideal' parameter, screen sharing will fail.

youennf opened the pull request:
#### 882a38ecdaf1712c1415cdbbe39cdb6c075eb021
<pre>
In M1 Ventura safari 16.1, when the constraints passed in by getDisplayMedia contain the &apos;ideal&apos; parameter, screen sharing will fail.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247310">https://bugs.webkit.org/show_bug.cgi?id=247310</a>
rdar://problem/101819966

Reviewed by Eric Carlson.

The issue is that the intrinsic size computation is wrong in GPUProcess.
This leads to a max size of (0, 0), which is used in case ideal constraints are used.

Instead of using CGDisplayCopyDisplayMode, we use getShareableContentWithCompletionHandler API which does not need any change to the sandbox.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::intrinsicSize const):

Canonical link: <a href="https://commits.webkit.org/256458@main">https://commits.webkit.org/256458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ae9da681b385592e95c57f81f735d09eb68a6fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5035 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28825 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105361 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165669 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5119 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33793 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88167 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101191 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101443 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3766 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82394 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30820 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87538 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73651 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39526 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37217 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20398 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4460 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41277 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43032 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43323 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39650 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -765,7 +765,7 @@
         "com.apple.ViewBridgeAuxiliary"
         "com.apple.coreservices.launchservicesd"
-        "com.apple.windowserver.active"))
+        "com.apple.coreservices.launchservicesd"))
```
Are there two entries for`com.apple.coreservices.launchservicesd` now?

pvollan commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -973,6 +973,7 @@
             "com.apple.replaykit.sharingsession"
             "com.apple.tccd.system"
             "com.apple.replayd"
+            "com.apple.windowserver.active"
```
It would be great if we could resolve this issue without allowing this sandbox extension. Is there another approach we could take? If this is a regression, perhaps we could track down the regression point?

youennf replied:
Done with API @eric-carlson mentionned

eric-carlson reviewed (approved):
r=me

youennf committed:
In M1 Ventura safari 16.1, when the constraints passed in by getDisplayMedia contain the 'ideal' parameter, screen sharing will fail.
https://bugs.webkit.org/show_bug.cgi?id=247310
rdar://problem/101819966

Reviewed by Eric Carlson.

The issue is that the intrinsic size computation is wrong in GPUProcess.
This leads to a max size of (0, 0), which is used in case ideal constraints are used.

Instead of using CGDisplayCopyDisplayMode, we use getShareableContentWithCompletionHandler API which does not need any change to the sandbox.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::intrinsicSize const):

Canonical link: https://commits.webkit.org/256458@main
